shared/pubsub.js

```diff
@@ -0,0 +1,386 @@
+'use strict'
+
+import type { JSONObject, JSONType } from '~/shared/types.js'
+
+// ====== Types ====== //
+
+export type Message = {
+  [key: string]: JSONType,
+  +type: string
+}
+
+export type PubSubClient = {
+  connectionTimeoutID: TimeoutID | void,
+  +customHandlers: Object,
+  failedAttempts: number,
+  isReconnecting: boolean,
+  +listeners: Object,
+  +messageHandlers: Object,
+  +options: Object,
+  +pendingSubscriptionSet: Set<string>,
+  +pendingUnsubscriptionSet: Set<string>,
+  reconnectionDelayID: TimeoutID | void,
+  socket: WebSocket | null,
+  +subscriptionSet: Set<string>,
+  +url: string,
+  // Methods
+  connect(): void,
+  destroy(): void,
+  getNextRandomReconnectionDelay(): number,
+  pub(contractID: string, data: JSONType): void,
+  sub(contractID: string): void,
+  unsub(contractID: string): void
+}
+
+export type SubMessage = {
+  [key: string]: JSONType,
+  +type: 'sub',
+  +contractID: string
+}
+
+export type UnsubMessage = {
+  [key: string]: JSONType,
+  +type: 'unsub',
+  +contractID: string
+}
+
+// ====== Enums ====== //
+
+export const NOTIFICATION_TYPE = Object.freeze({
+  ENTRY: 'entry',
+  PUB: 'pub',
+  SUB: 'sub',
+  UNSUB: 'unsub'
+})
+
+export const REQUEST_TYPE = Object.freeze({
+  PUB: 'pub',
+  SUB: 'sub',
+  UNSUB: 'unsub'
+})
+
+export const RESPONSE_TYPE = Object.freeze({
+  ERROR: 'error',
+  SUCCESS: 'success'
+})
+
+// ====== API ====== //
+
+/**
+ * Creates a pubsub client instance.
+ *
+ * @param {string} url - A WebSocket URL to connect to.
+ * @param {Object?} options
+ * {number?} timeout - Connection timeout duration in milliseconds.
+ * {boolean?} debug
+ * {object?} handlers - Custom handlers for socket events.
+ * {object?} messageHandlers - Custom handlers for different message types.
+ * {boolean?} reconnectOnDisconnection - Whether to reconnect after a server-side disconnection.
+ * {boolean?} reconnectOnOnline - Whether to reconnect after regaining internet connectivity.
+ * {boolean?} reconnectOnTimeout - Whether to reconnect after a timeout.
+ * @returns {Object}
+ */
+export function createClient (url: string, options?: Object = {}): PubSubClient {
+  const client = {
+    customHandlers: options.handlers || {},
+    failedAttempts: 0,
+    isReconnecting: false,
+    listeners: Object.create(null),
+    messageHandlers: { ...defaultMessageHandlers, ...options.messageHandlers },
+    options: { ...defaultOptions, ...options },
+    pendingSubscriptionSet: new Set(),
+    pendingUnsubscriptionSet: new Set(),
+    reconnectionDelayID: undefined,
+    socket: null,
+    subscriptionSet: new Set(),
+    connectionTimeoutID: undefined,
+    url: url.replace(/^http/, 'ws'),
```

Comment:
What about https/wss? Should that be handled here too? AFAIK that's how most servers will be setup.

Reply:
The second "s", if present, will not be affected by this prefix replacement:
```js
'https://example.com'.replace(/^http/, 'ws') // 'wss://example.com'
```

Reply:
Ah, perfect. Marking this as resolved.